### ISSUE TITLE
Combining statement/state space coverage from multiple runs

### ISSUE THREAD

**Issue by ajdavis** [tlaplus/tlaplus#423]
Related to model-based trace-checking (the same project as #413), I'm curious if TLC has a feature for combining coverage reports from multiple runs. This would be the scenario: We deploy trace-checking to our continuous integration server, and we want to accumulate coverage statistics over multiple executions of the trace-checker. The goal is to determine the total amount of coverage achieved by all execution traces for a given version of the specification and the actual software.

Since specifications usually have a small number of statements and a very large number of states, I imagine it would be easier to combine statement coverage reports than state space coverage reports.

Note: Since our project is paused, this is *not* a high-priority request. I'm filing it for two reasons. First, because I think it might be useful to some people. And second, if the feature is already implemented I want the opportunity to learn how to use it.

**Comment by lemmy** [comment]
Would a script do that combines TLC's coverage output?  It essentially has to find matching lines and increment the corresponding counters.

```
The coverage statistics at 2020-01-23 12:00:00
<Init line 34, col 1 to line 34, col 4 of module TraceWithIO>: 2:2
  line 34, col 9 to line 34, col 21 of module TraceWithIO: 2
```
```
The coverage statistics at 2020-01-24 12:00:00
<Init line 34, col 1 to line 34, col 4 of module TraceWithIO>: 2:1
  line 34, col 9 to line 34, col 21 of module TraceWithIO: 3
<Next line 35, col 1 to line 35, col 4 of module TraceWithIO (35 12 35 51)>: 3:6
```

Combined into:

```
The coverage statistics at 2020-01-25 12:00:00
<Init line 34, col 1 to line 34, col 4 of module TraceWithIO>: 2:3
  line 34, col 9 to line 34, col 21 of module TraceWithIO: 5
<Next line 35, col 1 to line 35, col 4 of module TraceWithIO (35 12 35 51)>: 3:6
```

**Comment by ajdavis** [comment]
Yes, that would work nicely. Could its output be visualized in the TLA+ Toolbox?

**Comment by lemmy** [comment]
Sure, just modify the coverage statistics block above in the MC.out file of an existing Toolbox model. Start the Toolbox and open the model's coverage editor.

**Comment by ajdavis** [comment]
Excellent. What do you think about combining state space coverage reports?

**Comment by lemmy** [comment]
That is part of the coverage statistics block (the 3:6 after Next above indicates that this action found 6 states three of which were distinct).   :-)

**Comment by ajdavis** [comment]
Very cool. So let's say that two distinct runs both print:
```
<Next line 35, col 1 to line 35, col 4 of module TraceWithIO (35 12 35 51)>: 3:6
```
Is there a way currently to know whether they covered the *same* 3 states, different ones, or a partially overlapping set of states? I'd need to know that in order to accumulate state space coverage. I suspect that TLC would have to save to disk whatever data structure it uses to store the state fingerprints it has seen; then we could union two saved fingerprint files to calculate the total state space coverage of two runs.

**Comment by lemmy** [comment]
For small state spaces you can "dump" the complete state graph in either a TLC specific or in dot/graphviz notation. I would assume that there exists tools to diff two digraphs in dot notation (or use basic text diff).  At any rate, what kind of questions do you hope this will answer?

**Comment by ajdavis** [comment]
I hope it would answer questions about how much state space has been covered by model-based trace-checking. Perhaps a team could use this to expand the number or kinds of tests they run with trace logging, in order to increase coverage. By exploring the uncovered portions of the graph in some way, they might discover testing gaps. Does that make sense?

In any case, you've answered my questions about TLC's current feature set. Please feel free to close this ticket, or to keep it open as a feature request, depending on your judgment. I'm not actually requesting any feature development for myself. Thanks!

**Comment by lemmy** [comment]
If this gets implemented, it is probably easiest to reuse the parsing already available in the Toolbox.

https://github.com/tlaplus/tlaplus/blob/master/org.lamport.tla.toolbox.tool.tlc.ui/src/org/lamport/tla/toolbox/tool/tlc/output/data/ActionInformationItem.java
https://github.com/tlaplus/tlaplus/blob/master/org.lamport.tla.toolbox.tool.tlc.ui/src/org/lamport/tla/toolbox/tool/tlc/output/data/CoverageInformationItem.java
...
